### PR TITLE
Add timeout in clients

### PR DESCRIPTION
**Pull request by romainreuillon**
Here is the code to make it possible to set a timeout value in xhr.

Any comment or improvement to make to make it in the repo?

Best,
Romain

### CODE REVIEW

File: xhr/client/src/main/scala/endpoints4s/xhr/EndpointsSettings.scala [hunk 1]
```diff
@@ -3,11 +3,12 @@ package endpoints4s.xhr
 import endpoints4s.Hashing
 
 import scala.annotation.nowarn
+import scala.concurrent.duration.{FiniteDuration}
 
 /** Settings for XHR interpreter.
   * @param baseUri Base of the URI of the service that implements the endpoints, can be absolute or relative (e.g. "http://foo.com" or "/bar")
   */
-final class EndpointsSettings private (val baseUri: Option[String]) extends Serializable {
+final class EndpointsSettings private (val baseUri: Option[String], val timeout: Option[FiniteDuration]) extends Serializable {
```
Comment (julienrf) [inline xhr/client/src/main/scala/endpoints4s/xhr/EndpointsSettings.scala:11]: Could you please also update the implementation of `hashCode`, `equals`, and `toString` to take into account the `timeout`?

**Review by julienrf — COMMENTED** [review]
Hey @romainreuillon, thank you for this contribution!

Would it make sense to add it to the `fetch` interpreter as well? Maybe @ptrdom would be interested to do it?

**Comment by ptrdom** [comment]
@romainreuillon Thank you for also updating `fetch` interpreter. This change would also benefit from tests, I think it should be possible to set up a relevant interaction with `stub-server`.

File: fetch/client/src/main/scala/endpoints4s/fetch/EndpointsSettings.scala [hunk 2]
```diff
@@ -3,38 +3,45 @@ package endpoints4s.fetch
...
 
   override def equals(other: Any): Boolean =
     other match {
       case that: EndpointsSettings =>
-        baseUri == that.baseUri
+        baseUri == that.baseUri && timeout == that.timeout
       case _ => false
     }
 
   override def hashCode(): Int = Hashing.hash(baseUri)
```
Comment (julienrf) [inline fetch/client/src/main/scala/endpoints4s/fetch/EndpointsSettings.scala:23]: You need to add the timeout here as well

**Review by julienrf — COMMENTED** [review]
Thank you for the update @romainreuillon!

As @ptrdom mentioned, would it be possible to add a test? You could create an endpoint in the stub server that would hang for some time (e.g. 2 sec) before returning a response. You could then check what happens with a longer or shorter timeout.

File: xhr/client/src/main/scala/endpoints4s/xhr/EndpointsSettings.scala [hunk 3]
```diff
@@ -3,38 +3,45 @@ package endpoints4s.xhr
...
       case that: EndpointsSettings =>
-        baseUri == that.baseUri
+        baseUri == that.baseUri && timeout == that.timeout
       case _ => false
     }
 
-  override def hashCode(): Int = Hashing.hash(baseUri)
+  override def hashCode(): Int = Hashing.hash(baseUri, timeout)
 
   @nowarn("cat=unused")
```
Comment (julienrf) [inline xhr/client/src/main/scala/endpoints4s/xhr/EndpointsSettings.scala:25]: It seems this annotation is not necessary?

**Comment by romainreuillon** [comment]
@julienrf Thanks for the guideline.

I'm gonna try to implement a few test. I am trying to understand how the test suite works and where the endpoints used by the tests are being implemented.

File: fetch/client/src/main/scala/endpoints4s/fetch/EndpointsSettings.scala (hunk 2 above)
Comment (romainreuillon) [reply fetch/client/src/main/scala/endpoints4s/fetch/EndpointsSettings.scala]: Good catch, I just added it

File: xhr/client/src/main/scala/endpoints4s/xhr/EndpointsSettings.scala (hunk 3 above)
Comment (romainreuillon) [reply xhr/client/src/main/scala/endpoints4s/xhr/EndpointsSettings.scala:25]: The annotation was in the code already. I don't know, which warning it solves.

Should I remove it?

**Comment by romainreuillon** [comment]
I implemented a unit test for the timeout. I implemented the timeout in http4s client, fetch and xhr. For these client the test passes.

I didn't find how to implement it in akka-client in a timely manner, could you give me a hand on this one?

Are there other clients that I have missed?

**Comment by ptrdom** [comment]
Of course you can also implement timeouts for akka-http client, but then you would need to look into how to implement it, probably this page is relevant - https://doc.akka.io/docs/akka-http/current/common/timeouts.html.

**Comment by ptrdom** [comment]
Also you can see in CI logs that `sttp-client` has also failed, so you need to implement timeouts for it too. Otherwise you could just create a new suite with its new endpoint for timeout testing and only have it implemented by specific interpreters you have implemented the timeouts for. Now your tests are in the main suite, which is implemented by all interpreters, even those you currently have not modified.

**Comment by romainreuillon** [comment]
sttp-client timeout is now implemented and all test are okay on my computer

**Comment by romainreuillon** [comment]
I am unsure on how the modification I made triggered this test errors. It seems related to some akka server config... Any idea?

**Comment by ptrdom** [comment]
The errors I see for 2.13 workflow seem to be related to `@nowarn` annotations.

**Comment by romainreuillon** [comment]
The @nowarn were in the code already. Should I remove them (as it seems unrelated to the timeout implem) ?

**Comment by julienrf** [comment]
Hey @romainreuillon, thank you so much for addressing all the points we’ve made. Yes, it’s true that the `@nowarn` annotations were there before your changes. I don’t remember why we introduced them, to be honest. It seems that only now they create these compilation errors, so could you try to remove them?

**Comment by romainreuillon** [comment]
I've just removed them

**Comment by romainreuillon** [comment]
Anything left preventing to merge this PR?

File: algebras/testkit/src/main/scala/endpoints4s/algebra/client/EndpointsTestSuite.scala [hunk 4]
```diff
@@ -13,7 +13,7 @@ trait ClientEndpointsTestApi extends EndpointsTestApi {
 
 trait EndpointsTestSuite[T <: ClientEndpointsTestApi] extends ClientTestBase[T] {
 
-  def clientTestSuite() = {
+  def clientTestSuite(timeoutTest: Boolean = true) = {
```
[the remark this answers is not in the document]
Comment (romainreuillon) [reply algebras/testkit/src/main/scala/endpoints4s/algebra/client/EndpointsTestSuite.scala:16]: The test is not supported by akka-http yet

[the remark this answers is not in the document]
Comment (romainreuillon) [reply algebras/testkit/src/main/scala/endpoints4s/algebra/client/EndpointsTestSuite.scala:16]: The difference between the ChunkedJsonEntities tests and the Client tests is that the Client tests are encapsulated in a method, whereas the ChunkedJsonEntities tests are implement directly within the trait.

Making this test modular using cake pattern would require to inject a new method and calling it in each test suite supporting the timeout settings. Would that be a solution or should we keep the boolean parameter until all clients implement the timeout setting?

[the remark this answers is not in the document]
Comment (julienrf) [reply algebras/testkit/src/main/scala/endpoints4s/algebra/client/EndpointsTestSuite.scala:16]: I don’t think we need the extra method in the trait (even that’s what in most of the existing test suites, I am not sure this provides value). I recommend doing something like this:

https://github.com/endpoints4s/endpoints4s/blob/3545ad7c118b73bf6f44cc0fefd1b08b2804953d/algebras/testkit/src/main/scala/endpoints4s/algebra/client/ChunkedEntitiesTestSuite.scala#L12

[the remark this answers is not in the document]
Comment (julienrf) [reply algebras/testkit/src/main/scala/endpoints4s/algebra/client/EndpointsTestSuite.scala:16]: I would prefer if you could move the test to a separate trait that you would mix only to the relevant interpreters (instead of using the boolean param). Would you mind making that change?

File: fetch/client/src/main/scala/endpoints4s/fetch/EndpointsSettings.scala [hunk 5]
```diff
@@ -2,39 +2,44 @@ package endpoints4s.fetch
...
 import endpoints4s.Hashing
 
-import scala.annotation.nowarn
+import scala.concurrent.duration.FiniteDuration
 
 /** Settings for XHR interpreter.
   * @param baseUri Base of the URI of the service that implements the endpoints, can be absolute or relative (e.g. "http://foo.com" or "/bar")
   */
-final class EndpointsSettings private (val baseUri: Option[String]) extends Serializable {
+final class EndpointsSettings private (val baseUri: Option[String], val timeout: Option[FiniteDuration]) extends Serializable {
```
Comment (julienrf) [inline fetch/client/src/main/scala/endpoints4s/fetch/EndpointsSettings.scala:10]: To keep the backward binary compatibility, you need to add back the previous constructor as a secondary constructor:

```suggestion
final class EndpointsSettings private (val baseUri: Option[String], val timeout: Option[FiniteDuration]) extends Serializable {

  def this(baseUri: Option[String]) = this(baseUri, None)
```

File: http4s/client/js/src/test/scala/endpoints4s/http4s/client/Http4sClientEndpointsJsonSchemaTest.scala [hunk 6]
```diff
@@ -48,7 +50,7 @@ class Http4sClientEndpointsJsonSchemaTest
       port = Some(stubServerPort)
     ),
     Uri.Scheme.http,
-    FetchClientBuilder[IO].create
+    FetchClientBuilder[IO].withRequestTimeout(FiniteDuration(2, TimeUnit.SECONDS)).create
```
Comment (julienrf) [inline http4s/client/js/src/test/scala/endpoints4s/http4s/client/Http4sClientEndpointsJsonSchemaTest.scala:53]: Why is this change necessary?

File: sttp/client/src/main/scala/endpoints4s/sttp/client/Endpoints.scala [hunk 7]
```diff
@@ -37,7 +21,8 @@ import _root_.sttp.client3.{
   */
 class Endpoints[R[_]](
     val host: String,
-    val backend: SttpBackend[R, Any]
+    val backend: SttpBackend[R, Any],
+    val timeout: Option[FiniteDuration] = None
```
Comment (julienrf) [inline sttp/client/src/main/scala/endpoints4s/sttp/client/Endpoints.scala:25]: You will probably have to re-introduce the old constructor signature as a secondary constructor here as well.

File: sttp/client/src/main/scala/endpoints4s/sttp/client/Endpoints.scala [hunk 8]
```diff
@@ -332,11 +323,15 @@ trait EndpointsWithCustomErrors[R[_]]
       //#endpoint-type
       {
     def apply(a: A): R[B] = {
-      val result = backend.send(request(a).response(asStringAlways))
-      backend.responseMonad.flatMap(result) { sttpResponse =>
-        decodeResponse(response, sttpResponse)
+        val result = backend.send(request(a).response(asStringAlways))
+        val responseFuture = backend.responseMonad.handleError(result) {
+          case e if e != null && e.getCause.getClass.getName.toLowerCase.contains("timeout") =>
```
Comment (julienrf) [inline sttp/client/src/main/scala/endpoints4s/sttp/client/Endpoints.scala:328]: Would it be possible to pattern match on the cause using a “typed pattern” instead of manipulating the class name? Something like:

~~~ scala
e.getCause match {
  case timeout: TimeoutException => ...
}
~~~

**Review by julienrf — CHANGES_REQUESTED** [review]
Thank you again Romain for your patience! Overall, I am in favor of merging your contribution, but there are still a couple of points that I would like to see addressed.

File: sttp/client/src/main/scala/endpoints4s/sttp/client/Endpoints.scala (hunk 7 above)
Comment (romainreuillon) [reply sttp/client/src/main/scala/endpoints4s/sttp/client/Endpoints.scala]: It has been implemented

**Comment by romainreuillon** [comment]
Dear @julienrf, 

thank you for having taken the time to review this PR. I addressed the issue you raised about this PR.

best,
Romain

File: http4s/client/js/src/test/scala/endpoints4s/http4s/client/Http4sClientEndpointsJsonSchemaTest.scala (hunk 6 above)
Comment (julienrf) [reply http4s/client/js/src/test/scala/endpoints4s/http4s/client/Http4sClientEndpointsJsonSchemaTest.scala:54]: Oh, I see, there is no need to introduce the timeout configuration in the Http4s interpreter trait (like you did with the fetch interpreter, for instance), because the Http4s client can already be configured with a timeout by the users. That’s great!

I wonder if we should follow the same approach with the sttp interpreter then? Let the users configure the timeout in the underlying backend? Would that be possible?

Comment (romainreuillon) [reply http4s/client/js/src/test/scala/endpoints4s/http4s/client/Http4sClientEndpointsJsonSchemaTest.scala:54]: I extracted the config from the mixin in sttp tests, is that what you meant?

**Comment by romainreuillon** [comment]
@julienrf I'm back from vacation. I saw that you requested one change concerning sttp tests. Is that all what is left before merging or did I miss some other points?

File: http4s/client/js/src/test/scala/endpoints4s/http4s/client/Http4sClientEndpointsJsonSchemaTest.scala (hunk 6 above)
Comment (julienrf) [reply http4s/client/js/src/test/scala/endpoints4s/http4s/client/Http4sClientEndpointsJsonSchemaTest.scala:54]: Hey Romain ! No, I am sorry, I wasn't clear. What I meant is that maybe instead of introducing a timeout setting in the sttp interpreter it would be possible to do nothing about the timeout in endpoints4s and let sttp handle it itself. Just like you did here with this http4s interpreter: you didn't change anything, but you simply configured the timeout in http4s itself.

Comment (romainreuillon) [reply http4s/client/js/src/test/scala/endpoints4s/http4s/client/Http4sClientEndpointsJsonSchemaTest.scala:54]: Okay I understand. I think this was the first thing I tried.

I just tried it again, using this config:

```scala
val backend = TryHttpURLConnectionBackend(options = SttpBackendOptions.connectionTimeout(FiniteDuration(2, TimeUnit.SECONDS)))
```

but unfortunately, the interpreter doesn't timeout

Comment (romainreuillon) [reply http4s/client/js/src/test/scala/endpoints4s/http4s/client/Http4sClientEndpointsJsonSchemaTest.scala:54]: I think that the global config it made for the connection timeout (at tcp level) and the other for the read.

It is not the same timeout:
https://sttp.softwaremill.com/en/latest/conf/timeouts.html

Comment (romainreuillon) [reply http4s/client/js/src/test/scala/endpoints4s/http4s/client/Http4sClientEndpointsJsonSchemaTest.scala:54]: Okay, I think I've actually found how to it.

I just committed it.

**Review by julienrf — APPROVED** [review]

**Comment by julienrf** [comment]
Thank you so much Romain for your work! I’ve pushed an additional commit to my fork to resolve the binary compatibility issues (hopefully). Do you mind fast-forward-merging it to your fork?

https://github.com/openmole/endpoints4s/compare/timeout...julienrf:endpoints:timeout

**Comment by julienrf** [comment]
I just fixed the code style, please merge my last commit again: https://github.com/openmole/endpoints4s/compare/timeout...julienrf:endpoints:timeout

**Comment by julienrf** [comment]
Thank you Romain, this has been released in 1.8.0
